### PR TITLE
Support comments containing images

### PR DESCRIPTION
**Pull request by es94129**
Finished the Week 3 Blobstore walkthrough.

- Modify interface
   - support file upload
   - `` to display comments with images
- Fetch Blobstore url and handle with `DataServlet`

The portfolio is now deployed [here](http://ychen-sps-summer20.appspot.com/index.html).

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -13,17 +13,26 @@
 // limitations under the License.
 
 package com.google.sps.servlets;
-
+import com.google.appengine.api.blobstore.BlobInfo;
```
Comment (ivancjw) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:16]: Pls add a blank line above

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.google.sps.servlets;
+
+class Comment {
+    String message;
+    String imageUrl;
```
Comment (ivancjw) [inline portfolio/src/main/java/com/google/sps/servlets/Comment.java:5]: private String message;
private String imageUrl;

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+    Comment(String message, String imageUrl) {
+        this.message = message;
+        this.imageUrl = imageUrl;
+    }
+}
```
Comment (ivancjw) [inline portfolio/src/main/java/com/google/sps/servlets/Comment.java:11]: Pls add a new line to the end of file

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -39,18 +48,23 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
-    ArrayList<String> messages = new ArrayList<>();
+    // ArrayList<String> messages = new ArrayList<>();
```
Comment (ivancjw) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:51]: Pls don't leave temp code in the pull request. If you intend to keep it, pls add some comments to state why you comment out the code.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -39,18 +48,23 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    ArrayList<Comment> comments = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
-      messages.add((String) entity.getProperty("message"));
+    //   messages.add((String) entity.getProperty("message"));
```
Comment (ivancjw) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:54]: Ditto

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (es94129) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:51]: Thanks for pointing out. I will check this more carefully before pull request next time.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -69,7 +83,44 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
```
Comment (ivancjw) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:102]: I think blobKey is from the request, which is user input, so it doesn't necessarily contain exactly one element. You'd better check this above, like return null or throw an exception.

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -66,18 +66,44 @@ setTimeout(function() {
...
 }
 
-function createListElement(text) {
+function createListElement(comment) {
     const liElement = document.createElement('li');
-    liElement.innerText = text;
+    liElement.innerText = comment['message'];
+
+    if (comment.hasOwnProperty('imageUrl')) {
+        var brElement = document.createElement('br');
```
Comment (ivancjw) [inline portfolio/src/main/webapp/script.js:84]: why this is a var instead of const?

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -66,18 +66,44 @@ setTimeout(function() {
...
+        liElement.appendChild(brElement);
```
Comment (ivancjw) [inline portfolio/src/main/webapp/script.js:85]: I think you can write it in one line:

liElement.appendChild(document.createElement('br'));

File: portfolio/src/main/webapp/index.html [hunk 9]
```diff
@@ -118,19 +118,22 @@ <h2>My life in Tsinghua</h2>
...
             <section>
                 <h2>leave a comment</h2>
                 <p>
-                    <form action="/data" method="POST">
-                        <input type="text" placeholder="name" name="name">
-                        <input type="email" placeholder="email" name="email"><br>
+                    <form id="comment-form" class="hidden" method="POST" enctype="multipart/form-data">
+                        <label class="btn_two">
+                            <input name="upload_img" style="display:none;" type="file"
+                            onchange="showUploadFilename(this.files[0].name)">
```
Comment (ivancjw) [inline portfolio/src/main/webapp/index.html:130]: text alignment:

```html
                            <input name="upload_img" style="display:none;" type="file"
                                   onchange="showUploadFilename(this.files[0].name)">
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (es94129) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:102]: Yeah this was also assured from the html code, which is not a good practice. I added a check above.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -69,7 +83,44 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     }
   }
 
-  static private String convertToJsonUsingGson(ArrayList<String> messages) {
-    return new Gson().toJson(messages);
+  static private String convertToJsonUsingGson(ArrayList<Comment> comments) {
+    return new Gson().toJson(comments);
+  }
+  
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
```
Comment (ivancjw) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:90]: Format: 
```
/**
 * One sentence description opens with descriptive verbs in the indicative mood.
 *
 * Detailed comments, if any.
 */
```

```java
/**
 * Returns a URL that points to the uploaded file, or null if the user didn't upload a file.
 */
```

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -66,18 +66,43 @@ setTimeout(function() {
...
+
+function showUploadFilename(filename) {
+    $('#upload-filename').html(filename);
+}
+
+function fetchBlobstoreUrlAndShowForm() {
+  fetch('/blobstore-upload-url')
+      .then((response) => {
+        return response.text();
+      })
```
Comment (ivancjw) [inline portfolio/src/main/webapp/script.js:102]: .then(response => response.text())

**Pull request merged**